### ISSUE TITLE
RKE2 Provisioning: Cluster Options - Cloud Provider Support

### ISSUE THREAD

**Issue by deniseschannon** [rancher/rancher#32390]
Ability to specify 

AWS
Azure
~Custom~
"external" (to be tested in https://github.com/rancher/rancher/issues/33611)

This is available in the agentArgs (https://github.com/rancher/rancher/issues/32376) issue, but calling it out specifically to have it specially tested for these use cases.

**Comment by deniseschannon** [comment]
I've created a specific ticket for vsphere since it leverages separate helm charts vs. cloud provider config within the cluster

**Comment by davidnuzik** [comment]
Custom is not an option in the UI so will skip this. There is (None), aws, azure, gcp, vsphere, and external.

**Comment by davidnuzik** [comment]
**My checks FAILED**
 
**Reproduction Steps:**

Not required.
 
**Validation Steps:**

Rancher version: v2.6-head 4b6af95d7 8/9/21 8:46 am Pacific
Rancher cluster type: single-node docker
Docker version: 19.03
 
Downstream cluster type: rke2
Downstream K8s version: v1.21.3-rc4+rke2r2
 
Validation steps:
_**AWS Use Case - FAIL**_
1. Provision an RKE2 cluster with the AWS provider in Rancher. Choose "aws" as the Cloud Provider in the Cluster Configuration section in the UI. Be sure to specify a valid IAM role as well as this is required for the aws cloud provider.
2. The cloud provider "aws" is indeed set on the RKE2 node:
```
root@ip-REDACTED_IP:/etc/rancher/rke2/config.yaml.d# cat 50-rancher.yaml 
{
  "agent-token": "SECRET-REDACTED-TOKEN",
  "cloud-provider-name": "aws",
  "cni": "calico",
  "disable-kube-proxy": false,
  "etcd-expose-metrics": false,
  "etcd-snapshot-retention": 5,
  "etcd-snapshot-schedule-cron": "0 */5 * * *",
  "node-label": [
    "rke.cattle.io/machine=1a3a2bb4-9df3-41b1-92d4-0ad940264532"
  ],
  "protect-kernel-defaults": false,
  "token": "SECRET-REDACTED-TOKEN"
```

Setting up the AWS cloud provider on an rke1 node works as expected - this only doesn't work with RKE2. The cluster will never get out of "Provisioning" state:
![image](https://user-images.githubusercontent.com/45179589/128790971-b2fcaf92-8cfa-40f6-8e56-34a316384e74.png)

I filed a separate bug for this here: https://github.com/rancher/rancher/issues/34105

In the meantime, I'll test other cloud providers and leave a new comment once I have test results:
Azure Use Case - To Do
External Use Case - To Do

**Comment by deniseschannon** [comment]
@davidnuzik Are we sure this works in standalone RKE2 cause if Rancher is passing the info to RKE2, then is this actually a Rancher integration issue or is it a standalone RKE2 issue?

**Comment by davidnuzik** [comment]
Never-mind, only IAM profile has to be given. @deniseschannon

**Comment by davidnuzik** [comment]
So aws cloud provider support is indeed broken. Standalone it works - Max tested this @deniseschannon Rancher is not doing something right presumably. However, looking at what the UI passes to backend it looks good to me.

**Comment by davidnuzik** [comment]
Blocked from testing azure until https://github.com/rancher/rancher/issues/34112 is fixed (new rc)

**Comment by davidnuzik** [comment]
Actually moving To Test. This isn't something that is really "next up" until all the blocker bugs are fixed.

**Comment by davidnuzik** [comment]
https://github.com/rancher/rancher/issues/34105 is fixed (can test aws now).
Azure however, still waiting on https://github.com/rancher/dashboard/issues/3798

**Comment by davidnuzik** [comment]
I should be unblocked from testing Azure now and AWS is working - will pick this up now.
https://github.com/rancher/rancher/issues/33611 will be used for testing external cp.

**Comment by davidnuzik** [comment]
########################################################### 
**My checks PASSED for AWS, Azure I am blocked by https://github.com/rancher/rancher/issues/34367**
###########################################################
 
**Reproduction Steps:**

Not required.
 
**Validation Environment:**
Rancher version: v2.6.0-rc4 (`11d89464d`) (tagged 8/17/21 6:43pm according to Docker Hub)
Rancher cluster type: single-node docker
Docker version: 20.10
 
Downstream cluster type: rke2
Downstream K8s version: v1.21.3-rc6+rke2r2
 
**Validation steps:**
_**AWS Cloud Provider Use Case:**_

1. Create a multi-node RKE2 cluster with the AWS node driver in Rancher. Specify a suitable IAM role (one which has necessary permissions for aws in-tree cloud provider) and specify the "Amazon" cloud provider. I set up the following pools:
"etcd" pool with just etcd role - 3 machines
"cp" pool with just controlplane role - 2 machines
"worker" pool with just worker role - 3 machines
Note: Due to a [bug](https://github.com/rancher/rancher/issues/34300), use canal CNI instead of default calico CNI.

2. Test Cloud provider features. Ensure ec2 nodes and cloud provider objects get tagged correctly.
- PASS - Test Storage by creating a storage class and creating a PVC - check in aws console for the volume and that has proper tag.
![image](https://user-images.githubusercontent.com/45179589/129972535-50c0f676-3292-4547-a735-a28bb40cee0a.png)
![image](https://user-images.githubusercontent.com/45179589/129972603-2190a1b9-6ffc-4ff0-a4cc-01440c630d44.png)


 - PASS - Test Load Balancer by creating a workload DaemonSet such as Nginx and creating a service that uses L4 LB - check in aws console for the load balancer and that has proper tag.
![image](https://user-images.githubusercontent.com/45179589/129972662-91d4fa4b-ffea-4cf7-9035-5006cddb50e2.png)
![image](https://user-images.githubusercontent.com/45179589/129972703-7706142c-3baa-4ac4-9209-da0e68b4e64d.png)

No issues observed during testing aws cloud provider. I did notice that if we delete a cluster the cloud provider objects are not removed (volume and lb). It looks like this has been how this has always worked though when checking rke1 and also rke1 with Rancher 2.5.9. As such I created https://github.com/rancher/rancher/issues/34293 feature request recently. I think we should be cleaning up these objects upon cluster deletion so the user does not need to manually remove these objects prior to cluster deletion or manually remove in aws console post cluster deletion.

_**Azure Cloud Provider Use Case:**_
###########################################################
**_NOT DONE - I run into this bug: https://github.com/rancher/rancher/issues/34367_**
###########################################################

Additional Info:
**_Will wait until Azure is working before I can continue with Azure cloud provider use case._**

**Comment by deniseschannon** [comment]
Since we have validated AWS and external has been validated with a separate issue. We'll keep/use the Azure issue #34367 to track cloud provider support.

Based on the 2 use cases validated, we've validated that sending cloud provider info to RKE2 is working, which was the intent of this issue to begin with.

**Comment by davidnuzik** [comment]
The good news is I can bring up Azure clusters now when specify azure cloud provider.
The bad news is that I still cannot leverage Azure cloud provider features -- my understanding is this is not a release blocker and we'll work to fix asap for a 2.6.1 release.

I have been working with Hussein and we tried a cloud provider config like the following but this didn't work either:

```
{
    "cloud":"AzurePublicCloud",
    "tenantId": "secret_tenantId",
    "aadClientId": "secret_aadClientId",
    "aadClientSecret": "secret_aadClientSecret",
    "subscriptionId": "secret_subscriptionId",
    "resourceGroup": "dave-nuzik2-docker-machine",
    "location": "westus",
    "subnetName": "dave-nuzik2-docker-machine",
    "securityGroupName": "dave-nuzik2-rancher-managed-uxZ23R3B",
    "securityGroupResourceGroup": "dave-nuzik2-docker-machine",
    "vnetName": "dave-nuzik2-docker-machine-vnet",
    "vnetResourceGroup": "dave-nuzik2-docker-machine",
    "primaryAvailabilitySetName": "dave-nuzik2-docker-machine",
    "routeTableResourceGroup": "dave-nuzik2-docker-machine",
    "cloudProviderBackoff": false,
    "useManagedIdentityExtension": false,
    "useInstanceMetadata": true
}
```

Still suspect that maybe our config is not quite right -- looking at azure documentation we should be good though.

**Comment by davidnuzik** [comment]
I need further assistance to move forward - am still unable to get Azure cp to work.

**Comment by davidnuzik** [comment]
Closing - this works now see https://github.com/rancher/rancher/issues/34367 for details. We still need to ensure rancher docs get updated - that issue will link to a docs issue and UI issue but this is functional today.
